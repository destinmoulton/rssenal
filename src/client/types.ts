import { List, Map, OrderedMap, Set } from "immutable";
import { Notification } from "react-notification-system";
/** Redux Interfaces */
export interface IDispatch {
    <R>(
        asyncAction: (dispatch: IDispatch, getState: () => IRootStoreState) => R
    ): R;
    <R>(asyncAction: (dispatch: IDispatch) => R): R;
    // (neverAction: (dispatch: Dispatch, getState: () => GetState) => never): never;
    (action: object): void;
    // (action: Thunk): ; // thunks in this app must return a promise
}

export interface IGetState {
    (): IRootStoreState;
}

export interface IAuthAction {
    type: string;
    username?: string;
    password?: string;
    message?: string;
}

export type TEntryID = string;

export interface IEntry {
    _id: TEntryID;
    feed_id: string;
    guid: string;
    title: string;
    link: string;
    creator: string;
    content: string;
    content_snippet: string;
    publish_date: string;
    has_read: boolean;
    feedTitle?: string;
    timeAgo?: string;
}

export interface IEntriesAction {
    type: string;
    entries?: TEntries;
}

export type TEntries = OrderedMap<TEntryID, IEntry>;

/** Feeds Interfaces & Types */

export type TFeedID = string;

export interface IFeed {
    _id: TFeedID;
    title: string;
    folder_id: string;
    description: string;
    link: string;
    creation_date: string;
    last_updated: string;
}

export type TFeeds = OrderedMap<TFeedID, IFeed>;

export interface IFeedsAction {
    type: string;
    feeds?: TFeeds;
    unreadMap?: IFeedsUnreadMap;
}

export type TUnreadMapFeeds = Map<TFeedID, number>;
export type TUnreadMapFolders = Map<TFolderID, number>;
export interface IFeedsUnreadMap {
    entriesCounted: Set<TFeedID>;
    feeds: TUnreadMapFeeds;
    folders: TUnreadMapFolders;
}

/** Filter Interfaces */

export interface IFilter {
    id?: string;
    limit: string;
}
export interface IFilterAction {
    type: string;
    newFilter?: object;
    filterTitle?: string;
    filteredEntries?: TEntries;
}

/** Folders Interfaces & Types */

export type TFolderID = string;

export interface IFolder {
    _id: TFolderID;
    name: string;
    order: number;
}

export type TFolders = OrderedMap<TFolderID, IFolder>;

export interface IFolderAction {
    type: string;
    folders?: TFolders;
}

/** Messages Interfaces */
export interface IMessage {
    message: string;
    level: "error" | "warning" | "info" | "success";
    uid?: number;
    onRemove?: (message: IMessage) => void;
    position?: "br" | "tr" | "tl" | "tc" | "bl" | "bc";
}

export type TMessages = List<IMessage>;

export interface IMessageAction {
    type: string;
    messages: List<IMessage>;
    lastUID?: number;
}

/** Redux Reducer Interfaces */

export interface IReducerStateAuth {
    authenticationError: string;
    isAuthorized: boolean;
    isValidatingToken: boolean;
}

export interface IReducerStateEntries {
    entries: TEntries;
}

export interface IReducerStateFeeds {
    feeds: TFeeds;
    unreadMap: IFeedsUnreadMap;
    isAddingFeed: boolean;
    isUpdatingFeed: boolean;
}

export interface IReducerStateSettings {
    settings: TSettings;
}

export interface IReducerStateFilter {
    filter: IFilter;
    filterTitle: string;
    filteredEntries: TEntries;
}

export interface IReducerStateFolders {
    folders: TFolders;
    hasFolders: boolean;
    isDeletingFolder: boolean;
    isFetchingFolders: boolean;
    isSavingFolder: boolean;
}

export interface IReducerStateMessages {
    messages: TMessages;
    lastUID: 0;
}

export interface IRootStoreState {
    authStore: IReducerStateAuth;
    entriesStore: IReducerStateEntries;
    feedsStore: IReducerStateFeeds;
    filterStore: IReducerStateFilter;
    foldersStore: IReducerStateFolders;
    messagesStore: IReducerStateMessages;
    settingsStore: IReducerStateSettings;
}

/** Settings Interfaces */

export interface ISetting {
    name: string;
    type: string;
    refresh_entries_on_change: boolean;
    value: any;
}

export type TSettings = Map<string, ISetting>;

export interface ISettingsAction {
    type: string;
    settings?: TSettings;
}
