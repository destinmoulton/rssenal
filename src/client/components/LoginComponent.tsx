import debug from "debug";
import * as React from "react";

import {
    Button,
    Form,
    Grid,
    Header,
    Icon,
    Message,
    Segment
} from "semantic-ui-react";

const log = debug("rssenal:LoginComponent");

export interface ILoginMapState {
    authenticationError: string;
}

export interface ILoginMapDispatch {
    authLoginUser: (username: string, password: string) => void;
}

type TAllProps = ILoginMapDispatch & ILoginMapState;

interface ILoginState {
    username?: string;
    password?: string;
}

class LoginComponent extends React.Component<TAllProps, ILoginState> {
    state = {
        username: "",
        password: ""
    };

    _changeInput = (e: React.FormEvent<HTMLInputElement>) => {
        this.setState({ [e.currentTarget.name]: e.currentTarget.value.trim() });
    };

    _authenticateUser = () => {
        log("_authenticateUser()");
        const { username, password } = this.state;

        if (username && password) {
            this.props.authLoginUser(username, password);
        }
    };

    render() {
        log("render()");
        const { authenticationError } = this.props;

        let message = null;
        if (authenticationError !== "") {
            message = (
                <Message negative>
                    <p>{authenticationError}</p>
                </Message>
            );
        }
        return (
            <div className="login-form">
                <style>{`
                body > div,
                body > div > div,
                body > div > div > div.login-form {
                    height: 100%;
                }
                `}</style>
                <Grid
                    textAlign="center"
                    style={{ height: "100%" }}
                    verticalAlign="middle"
                >
                    <Grid.Column style={{ maxWidth: 450 }}>
                        <Header as="h2" className="rss-login-header">
                            <Icon name="rss" />
                            &nbsp;&nbsp;RSSenal
                        </Header>
                        <Form size="large">
                            <Segment>
                                <Form.Input
                                    fluid
                                    icon="user"
                                    iconPosition="left"
                                    placeholder="Username"
                                    name="username"
                                    onChange={this._changeInput}
                                />
                                <Form.Input
                                    fluid
                                    icon="lock"
                                    iconPosition="left"
                                    placeholder="Password"
                                    type="password"
                                    name="password"
                                    onChange={this._changeInput}
                                />
                                {message}
                                <Button
                                    className="rss-login-button"
                                    fluid
                                    size="large"
                                    onClick={this._authenticateUser}
                                >
                                    Login
                                </Button>
                            </Segment>
                        </Form>
                    </Grid.Column>
                </Grid>
            </div>
        );
    }
}

export default LoginComponent;
