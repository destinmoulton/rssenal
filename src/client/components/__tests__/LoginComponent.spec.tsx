import * as React from "react";
import { shallow } from "enzyme";
import * as Types from "../../types";
import LoginComponent from "../LoginComponent";

describe("<LoginComponent />", () => {
    it("renders and matches snapshot", () => {
        const wrapper = shallow(
            <LoginComponent
                authenticationError={"Test Error"}
                loginUser={jest.fn()}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });
});
