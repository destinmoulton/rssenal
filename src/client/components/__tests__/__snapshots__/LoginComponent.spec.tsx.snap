// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginComponent /> renders and matches snapshot 1`] = `
<div
  className="login-form"
>
  <style>
    
                body &gt; div,
                body &gt; div &gt; div,
                body &gt; div &gt; div &gt; div.login-form {
                    height: 100%;
                }
                
  </style>
  <Grid
    style={
      Object {
        "height": "100%",
      }
    }
    textAlign="center"
    verticalAlign="middle"
  >
    <GridColumn
      style={
        Object {
          "maxWidth": 450,
        }
      }
    >
      <Header
        as="h2"
        className="rss-login-header"
      >
        <Icon
          as="i"
          name="rss"
        />
          RSSenal
      </Header>
      <Form
        as="form"
        size="large"
      >
        <Segment>
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            icon="user"
            iconPosition="left"
            name="username"
            onChange={[Function]}
            placeholder="Username"
          />
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            icon="lock"
            iconPosition="left"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
          />
          <Message
            negative={true}
          >
            <p>
              Test Error
            </p>
          </Message>
          <Button
            as="button"
            className="rss-login-button"
            fluid={true}
            onClick={[Function]}
            role="button"
            size="large"
          >
            Login
          </Button>
        </Segment>
      </Form>
    </GridColumn>
  </Grid>
</div>
`;
